Userland/Libraries/LibAudio/MP3Types.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2021, Arne Elster <arne@elster.li>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Array.h>
+#include <AK/EnumBits.h>
+
+namespace Audio::MP3 {
+
+enum class Mode {
+    Stereo,
+    JointStereo,
+    DualChannel,
+    SingleChannel,
+};
+
+enum class ModeExtension {
+    Stereo = 0,
+    IntensityStereo = 1,
+    MsStereo = 2,
+};
+AK_ENUM_BITWISE_OPERATORS(ModeExtension)
+
+enum class Emphasis {
+    None,
+    Microseconds_50_15,
+    Reserved,
+    CCITT_J17,
+};
+
+enum class BlockType {
+    Normal,
+    Start,
+    Short,
+    End,
+};
+
+struct Header {
+    i32 id;
+    i32 layer;
+    bool protection_bit;
+    i32 bitrate;
+    i32 sampling_frequency;
+    bool padding_bit;
+    bool private_bit;
+    Mode mode;
+    ModeExtension mode_extension;
+    bool copyright_bit;
+    bool original_bit;
+    Emphasis emphasis;
+    u16 crc16;
+    size_t frame_size;
+    size_t slot_count;
+
+    size_t channel_count() const { return mode == Mode::SingleChannel ? 1 : 2; }
+};
+
+struct Granule {
+    AK::Array<double, 576> samples;
+    AK::Array<AK::Array<double, 18>, 32> filter_bank_input;
+    AK::Array<AK::Array<double, 32>, 18> pcm;
+    u32 part_2_3_length;
+    u32 big_values;
+    u32 global_gain;
+    u32 scalefac_compress;
+    bool window_switching_flag;
+    BlockType block_type;
+    bool mixed_block_flag;
+    AK::Array<int, 3> table_select;
+    AK::Array<int, 3> sub_block_gain;
+    u32 region0_count;
+    u32 region1_count;
+    bool preflag;
+    bool scalefac_scale;
+    bool count1table_select;
+};
+
+struct Channel {
+    AK::Array<Granule, 2> granules;
+    AK::Array<int, 39> scale_factors;
+    AK::Array<int, 4> scale_factor_selection_info;
+};
+
+struct MP3Frame {
+    Header header;
+    AK::Array<Channel, 2> channels;
```

Comment:
As we may have only a single channel of audio, this needn't contain two channels. Use FixedArray with 1 or 2 elements instead.

Reply:
That would mean quite a few allocations. One FixedArray per frame and then one or two copies on top of that.

Reply:
AFAIK, copy semantics of FixedArray and Array are the same, so no, you shouldn't get extra copies. And yes, there are allocations, but you're allocating anyways each time you create the struct.